spring-integration-event/src/main/java/org/springframework/integration/event/inbound/ApplicationEventListeningMessageProducer.java

```diff
@@ -116,20 +129,45 @@ public void onApplicationEvent(ApplicationEvent event) {
 		}
 	}
 
+	@Override
+	protected Object evaluatePayloadExpression(Object rootObject) {
+		if (!this.payloadExpressionSet && rootObject instanceof PayloadApplicationEvent) {
+			return ((PayloadApplicationEvent) rootObject).getPayload();
+		}
+		else {
+			return super.evaluatePayloadExpression(rootObject);
+		}
+	}
+
 	private boolean stoppedRecently() {
 		return this.stoppedAt > System.currentTimeMillis() - 5000;
 	}
 
 	@Override
-	public boolean supportsEventType(Class<? extends ApplicationEvent> eventType) {
+	public boolean supportsEventType(ResolvableType eventType) {
 		if (this.eventTypes == null) {
 			return true;
 		}
-		for (Class<? extends ApplicationEvent> type : this.eventTypes) {
+
+		for (ResolvableType type : this.eventTypes) {
 			if (type.isAssignableFrom(eventType)) {
 				return true;
 			}
 		}
+
+		if (PayloadApplicationEvent.class.isAssignableFrom(eventType.getRawClass())) {
```

Comment:
@snicoll , would you mind elaborating a bit more here.
What don't you really like from this code?
Yours one in the `ApplicationListenerMethodAdapter` looks like:
````java
public boolean supportsEventType(ResolvableType eventType) {
	for (ResolvableType declaredEventType : this.declaredEventTypes) {
		if (declaredEventType.isAssignableFrom(eventType)) {
			return true;
		}
		else if (PayloadApplicationEvent.class.isAssignableFrom(eventType.getRawClass())) {
	        	ResolvableType payloadType = eventType.as(PayloadApplicationEvent.class).getGeneric();
			if (declaredEventType.isAssignableFrom(payloadType)) {
				return true;
			}
		}
	}
	return eventType.hasUnresolvableGenerics();
}
````
So, where is the difference? Unless my one gets deal with the list of supported event types.